server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -36,6 +40,25 @@
 
     public static final String BACKING_INDEX_PREFIX = ".ds-";
     public static final DateFormatter DATE_FORMATTER = DateFormatter.forPattern("uuuu.MM.dd");
+    // Datastreams' leaf readers should be sorted by desc order of their timestamp field, as it allows search time optimizations
+    public static Comparator<LeafReader> DATASTREAM_LEAF_READERS_SORTER =
+        Comparator.comparingLong(
+            (LeafReader r) -> {
+                try {
+                    PointValues points = r.getPointValues(DataStream.TimestampField.FIXED_TIMESTAMP_FIELD);
+                    if (points != null) {
+                        byte[] sortValue = points.getMaxPackedValue();
+                        return LongPoint.decodeDimension(sortValue, 0);
+                    }
+                } catch (IOException e) {
+                    throw new IllegalStateException("Failed reading [" +
+                        DataStream.TimestampField.FIXED_TIMESTAMP_FIELD + "] field for the data stream!");
+                }
+                // this may happen if the segment contains only deleted documents
+                assert(r.numDocs() == 0);
+                return Long.MIN_VALUE;
```

Comment:
@jimczi  It looks like we need to allow to have points to be `null` for cases when a segment contains only deleted documents.  Are you ok with the current implementation:

- Throw IllegalStateException if we can't access points
- Return `Long.MIN_VALUE` if a segment contains only deleted docs

Reply:
We can also check the number of docs upfront ?:
````
if (r.numDocs() == 0) {
  return Long.MIN_VALUE;
}
````

Reply:
Good suggestion, done in 4c5bddc72400e1a4af9da6292754ea7cf8c8ace4